README
Darwinbuild
Updated 2009-02-17

Kevin Van Vechten <kvv@apple.com>
William Siegrist <wsiegrist@apple.com>

1. Overview
  1.1 Availability, Bug Reports, Contributions, and Discussion
2. Installation
  2.1 Creating the Build Directory
3. Building Darwin Projects (darwinbuild)
  3.1 Install Headers
  3.2 Chroot Builds
4. Build Environment Variables
  4.1 Build Aliases and Alternate Targets
5. Tips and Techniques
  5.1 Private Headers
  5.2 Build Tools
  5.3 Static Libraries
  5.4 Alternate BuildRoot Storage
Appendix
A. darwinxref

===========
1. Overview
===========

Darwinbuild is a collection of tools that assist compilation of
the many projects contained in Darwin[1], the open source base of Apple's
Mac OS X operating system[2].  Apple publishes the sources of these projects
in an archive format (.tar.gz).  An archive is published for each project
version on Apple's site[3].

These tools will provide the proper build environment as well as help to
resolve any necessary dependencies prior to building.

[1] <http://developer.apple.com/darwin/>
[2] <http://www.apple.com/macosx/>
[3] <http://www.opensource.apple.com/>


1.1 Availability, Bug Reports, Contributions, and Discussion

Due to the constant development of darwinbuild, the only version you should 
be working with is trunk of the subversion repository[1]. Trunk should 
always be stable enough for building projects. If you find a problem or have
a enhancement in mind, file a ticket at the website[2].

Discussion about darwinbuild, or building Darwin projects in
general, should take place on the darwinbuild-dev[3] mailing list.

[1] <http://svn.macosforge.org/repository/darwinbuild/trunk>
[2] <http://darwinbuild.macosforge.org/report/1>
[3] <http://lists.macosforge.org/mailman/listinfo/darwinbuild-dev>

===============
2. Installation
===============

You can install the latest version of darwinbuild using MacPorts. 

  % port install darwinbuild

Or, you can also install by checking out the source and building 
darwinbuild yourself. The included Xcode project will compile the tools and install 
them into the location specified by the PREFIX environment parameter.  If the 
PREFIX parameter is not specified, the files will be installed into /usr/local.

By default, darwinbuild will build for x86_64 and i386. You can override this with
the ARCHS environment variable. However, building trunk for ppc is not supported
since trunk is targeted for Darwin 10. The release branches for Darwin 8 and 9 do
support building for ppc. 

To build and install darwinbuild to the boot partition:

  % cd <path to top of source>
  % sudo xcodebuild install DSTROOT=/

2.1 Creating the Build Directory

After installation, you must initialize the build directory using darwinbuild. 
Assuming you wanted to build projects from 9G55 in your home directory:

  # mkdir ~/9G55
  # cd ~/9G55
  # sudo -s
  # darwinbuild -init 9G55
  Creating build root disk image ...
  Attempting to download http://svn.macosforge.org/repository/darwinbuild/trunk/plists//9G55.plist ...
  Download complete
  Attempting to download http://svn.macosforge.org/repository/darwinbuild/trunk/plists//9F33.plist ...
  Download complete
  ...
  Attempting to download http://svn.macosforge.org/repository/darwinbuild/trunk/plists//9A581.plist ...
  Download complete
  Initialization Complete
  # ls
  .build  Headers Logs    Roots   Sources Symbols


After initialization, the build directory will contain the following:
.build     contains private data for the darwinbuild system
Headers    contains the resulting header files from previous builds
Logs       contains logs of previous build attempts 
Roots      contains the finished products of previous successful builds
Sources    contains sources downloaded from the web
Symbols    contains debug symbol versions of previous build products

When using the darwinbuild command, it is necessary that the current
working directory be this build directory, or alternatively,
that the DARWIN_BUILDROOT environment variable be set to the absolute
path of the destination directory.  The example usages in this document
assume the current working directory is this build directory.


=========================================
3. Building Darwin Projects (darwinbuild)
=========================================

To build a Darwin project, for example xnu, the darwinbuild script can be
used in the following manner:
  # darwinbuild xnu

The darwinxref tool is consulted to find the version that corresponds to the
build specified when the build directory was initialized.  It is necessary
to run the darwinbuild tool as root so that projects can set the proper
ownership and permissions on the build results.

darwinbuild first looks in the Sources directory for a directory containing
the sources to be built (Sources/xnu-517.11.1), or a .tar.gz archive
containing the sources (Sources/xnu-517.11.1.tar.gz).  If neither is found,
darwinbuild will attempt to download the sources from the web.

If it does not already exist, a BuildRoot directory will be created.  This
is where the build will actually take place.  During the build, darwinbuild
will change the root directory to BuildRoot (see the chroot(8) man page for
details).  Darwinbuild is capable of copying the required tools, libraries
and headers from the Roots directory into the BuildRoot prior to building.
If a necessary dependency is not found in the Roots directory, it will be
downloaded from the web.

The build output will be written to the console, and additionally logged
into a file in the Logs directory.  The above example produces the
following file:
  Logs/xnu/xnu-517.11.1.log~1

The ~1 indicates that this log file corresponds to the first attempt to
build xnu version 517.11.1.  Each subsequent attempt will add one to this
build number.

If the build succeeds, the finished product will be copied out of the
BuildRoot directory and into the Roots directory:
  Roots/xnu/xnu-517.11.1.root~1
After the copy, darwinbuild traverses the directory and records all files
found in the darwinxref database.  This makes it possible to query which
project a file is produced by.  When a Mach-O executable, library, or bundle
is found during the traversal, the dynamic library load commands are recorded
in the darwinxref database.  This makes it possible to query which additional
projects are required to run an executable produced by the project.

Additionally, any products containing debug symbols will be placed into
the Symbols directory.


3.1 Install Headers

Passing the -headers flag to darwinbuild will start an alternative build
style where only the project's headers are produced.  This is useful for
working around various circular dependencies while compiling low level
projects like xnu, cctools, Libc, etc.

The results of a -headers build are placed in the Headers directory, in
the same style as the Roots directory for a regular build.  For example:
  % sudo darwinbuild -headers IOKitUser
  % ls Headers/IOKitUser/IOKitUser-184.hdrs~1


3.3 Logging Dependencies

Passing the -logdeps flag to darwinbuild will build the specified project
while recording the paths of all files openened and all executables invoked.
These paths are written into the Logs directory, and if the build is
successful, will be imported into the darwinxref database.  


==============================
4. Build Environment Variables
==============================

At the minimum, the build environment consists of the creation of SRCROOT,
OBJROOT, SYMROOT, and DSTROOT environment variables.  These variables contain
an absolute path to a directory which must exist prior to invoking the Makefile.
User configurable environment variables, such as the target architecture,
can be set in the build plist file that is loaded via loadIndex.  The
darwinbuild script creates the aforementioned directories, sets the environment
variables, and issues the appropriate make command.

Within the plist file, the RC_ARCHS variable indicates which target
architectures should be included in the build.  Currently only ppc and i386
are available.  For each architecture in RC_ARCHS, and equivalent RC_{arch}
should be set to YES (i.e. RC_ARCHS="ppc" RC_ppc=YES).

The RC_NONARCH_CFLAGS specify additional command line flags to be passed to the C
compiler during the build.  On Darwin, the -no-cpp-precomp flag should be
passed since cpp-precomp is not available.  During the build, the RC_ARCHS and
RC_NONARCH_CFLAGS variables will be combined to create the RC_ARCHS environment
variable; a -arch flag will be appended for each architecture listed in RC_ARCHS.

The MACOSX_DEPLOYMENT_TARGET variable should match the major Mac OS X release
that is being targeted, such as 10.2 or 10.3.


4.1 Build Aliases and Alternate Targets

Some projects may produce different results based on the contents of 
the RC_ProjectName variable.  When the same source archive is used
to create more than one component of Darwin, it's referred to as a build
alias.  Because of this, it is important to always provide accurrate
project name and version information in the environment.  By default,
darwinbuild sets these variables appropriately.

By default, "install" (for non-headers builds) is the first argument passed
to the make tool (gnumake or xcodebuild).  However, some projects produce
different results based on alternate targets.  If the project in the property
list contains a "target" attribute, that string will be passed instead.


======================
5. Tips and Techniques
======================


5.1 Private Headers

Many open source header files are not present in a standard Mac OS X install.
These "private headers" are not needed by application developers, and are not
part of any SDK officially supported by Apple.  However, Darwin projects are
part of the operating system itself and these headers are required.  The
darwinbuildheaders command uses the darwinxref tool and darwinbuild -headers
commands to produce all header files for a given darwin release.
  % bin/darwinbuildheaders

The resulting header files from each project are placed into the Headers
directory.  The headers can be subsequently installed into the BuildRoot
directory using the installheaders command.
  % bin/installheaders

By modifying the build plist file, it is possible to include these headers
instead of the default system headers, giving more accurate build results and
avoiding failures.  Edit the RC_NONARCH_CFLAGS variable to provide the following
include paths to gcc (substituting the actual installation directory).
  -I/usr/local/darwinbuild/BuildRoot/usr/include
  -I/usr/local/darwinbuild/BuildRoot/usr/local/include


5.2 Build Tools

In addition to private headers, there are numerous command line tools which
are used exclusively to build Darwin projects.  These tools are placed in
/usr/local/bin.  For example, building xnu requires kextsymboltool,
relpath, and decomment.


5.3 Static Libraries

Several projects require static libraries from other projects.  These libraries
are usually placed in /usr/local/lib/system.  The most notable examples are
xnu's usage of libkld.a which is produced by cctools_ofiles, and Libsystem's
use of many static libraries from Libc, Libinfo, and more.


5.4 Alternate BuildRoot Storage

By default, "darwinbuild -init" will setup your environment to use a HFSX
sparsebundle to host your BuildRoot. A sparsebundle is used to work around
problems with using xcodebuild inside of a chroot. There are two other
storage methods to choose from: NFS Loopback and Legacy Directory. 

  1. NFS Loopback      (ex: darwinbuild -init 9G55 -nfs)
     The NFS loopback method places a directory under .build and adds that 
     directory to /etc/exports. When you try to build a project, darwinbuild 
     will mount the NFS export on localhost. The NFS mount also works around 
     the xcodebuild problems. 

  2. Legacy Directory  (ex: darwinbuild -init 9G55 -nodmg)
     This is just the way darwinbuild worked before the sparsebundle and NFS
     support was added. It just makes a directory for BuildRoot and builds 
     on whatever filesystem you are on at the time. If you are on HFS or ZFS,
     this storage will cause problems for Xcode-based projects. 
  

=============
A. darwinxref
=============

The darwinxref tool allows you to query which source version of a Darwin
project is present in a particular Mac OS X build.  It also stores information
about what dependencies a particular project has, and what files the project
produces.  As each Mac OS X release is made available, Apple publishes a
property list file containing the project names and versions in that release.
These property lists are read by the darwinxref tool to seed its internal
database.  Property lists are available via Subversion[1].

The darwinxref tool uses SQLite[2] to maintain its database of projects,
versions, files and dependencies.  This distribution includes a pre-built
version of the sqlite library in the file libsqlite3.a, and its associated
header file, sqlite3.h.

[1] <http://svn.macosforge.org/repository/darwinbuild/trunk/plists>
[2] <http://www.sqlite.org/>

An example of downloading, installing, and querying a property list:
  % curl http://svn.macosforge.org/repository/darwinbuild/trunk/plists/7U16.plist > \
      plists/7U16.plist
  % bin/darwinxref loadIndex plists/7U16.plist
  268 of 268 projects loaded.
  % bin/darwinxref -b 7U16 version xnu
  xnu-517.11.1

To list all projects in a build, use the special project name '*':
  % bin/darwinxref -b 7U16 version '*'

To register the results of a previous build with the database use the
register command.  Note this is done automatically by darwinbuild:
  % bin/darwinxref register adv_cmds 63 Roots/adv_cmds/adv_cmds-63.root~1
  /bin
  /bin/ps
  /bin/stty
  ...

To find which project produces the 'whois' command by searching the
list of previously registered files:
  % bin/darwinxref findFile whois
  adv_cmds-63:
          /usr/bin/whois




CHANGES
Darwin Build Scripts Change History
-----------------------------------
Release 37 [16-Oct-2013]
	- darwinup: fix C++ 98 conformance issue

Release 36 [2-May-2013]
	- darwinup: fix bad return code from update_xpc_services_cache.
	- darwinup: don't run bzip2 compression on (limited cpu) embedded devices.

Release 35 [24-Apr-2013]
	- all: fix implicit integer conversions.

Release 34.1 [14-Mar-2013]
Release 34 [12-Mar-2013]
	- darwintrace: avoid snprintf(3); manually construct log messages.

Release 33 [29-Aug-2012]
	- darwinup: Fix uninitialised variable in Database.cpp.

Release 32 [17-May-2012]
	- darwinup: Updated trigger for xpchelper cacher rebuild.

Release 31 [22-Mar-2012]
	- darwintrace: fixed crash when used by multi-threaded applications

Release 30 [15-Mar-2012]
	- darwinup: not detecting installs into /System/Library/Sandbox/Profiles

Release 29 [15-Feb-2012]
	- darwintrace: posix_spawn may have NULL envp parameter

Release 28.1 [9-Feb-2012]
	- darwinup: Don't include actual mach-o binaries in the test suite.

Release 28 [8-Feb-2012]
	- darwinup: Display more helpful error if unable to create a depot.
	- darwinup: Touch /System/Library/Extensions when uninstalling kernel extensions.
	- darwinup: Run xpchelper on system sandbox profile changes.

Release 27 [23-Jan-2012]
	- darwinup: Clear quarantine xattr from installed files.
	- darwinup: Correctness fix for has_suffix.

Release 26.5 [29-Nov-2011]
        - all: set correct INSTALL_PATH for iOS simulator

Release 26.4 [17-Nov-2011]
	- darwintrace: files opened with fopen(3) are missing from trace log

Release 26.3 [11-Nov-2011]
	- darwintrace: darwintrace_make_environ bug unsets DYLD_INSERT_LIBRARIES

Release 26.2 [9-Nov-2011]
	- darwintrace: build failure on 10.6, undefined symbol __posix_spawn
	- darwintrace: crash in darwintrace_make_environ

Release 26.1 [7-Nov-2011]
	- darwintrace: compiler warning build fix

Release 26 [6-Nov-2011]
	- darwintrace: preserve DYLD_INSERT_LIBRARIES and DARWINTRACE_LOG
	- darwintrace: interpose posix_spawn

Release 25 [4-Nov-2011]
	- darwinup: support for rebuilding xpchelper cache

Release 24.1 [17-Jun-2011]
	- Build fixes for newer compilers

Release 24 [10-Feb-2011]
	- darwinup: fix uninstall build-check safety
	- darwinup: remove unprebinding support
	- darwintrace: use modern interposing
	- darwintrace: add feature to ignore ROOT paths
	- darwintrace: add test suite

Release 23.1 [25-Jan-2011]
	- darwinup: strip symbols when building with Public configuration

Release 23 [03-Nov-2010]
	- darwinup: properly backup and restore existing content when
	  installing a root that replaces a directory with a file. 

Release 22 [27-Oct-2010]
	- darwinup: make man page clearer about what paths we support
	- darwinup: add safety measures to prevent recursive installs
	- darwinup: do not try to rollback installs that fail before
	  there is anything to rollback. 

Release 21.1 [29-Sep-2010]
	- darwinup: fix regression in v21 that broke installing roots when
	  the path is a symlink to a directory. 

Release 21 [17-Sep-2010]
	- darwinup: fix bug which silently failed to fully install
	  a root if the root had a directory where the destination
	  had a symlink to a directory
	- darwinup: test for object change during analyze stage instead
	  of install stage so we can bail out without having a root
	  partially installed then uninstalled
	- darwinup: add a full set of test cases for object change
	  conditions

Release 20.1 [02-Sep-2010]
	- darwinxref: build fix for format string warning

Release 20 [10-Aug-2010]
	- all: use CommonCrypto instead of OpenSSL
	- all: remove obsolete Makefiles

Release 19.1 [01-Jul-2010]
	- darwinup: fix bug which prevented remote root install when the remote
	  path is a symlink.

Release 19 [15-Jun-2010]
	- darwinup: fix crash when getgrnam fails
	- darwinup: add rename command
	- darwinup: touch /System/Library/Extensions when a kernel 
	  extension is modified.
	- darwinup: add option for rebooting when finished

Release 18.1 [04-May-2010]
	- darwinup: fix release build issue

Release 18 [03-May-2010]
	- darwinup: fix bug which caused the uninstall half of upgrade to
	  run when the install half failed
	- darwinup: add .tbz as a suffix for bzip2 tarballs

Release 17.1 [20-Apr-2010]
	- darwinup: add a manpage for darwinup
	- darwinup: fix crash when not root and a db upgrade is needed
	- darwinup: make remote root install behave consistently with
	  respect to trailing slashes.

Release 17 [30-Mar-2010]
	- darwinup: only update dyld cache when Depot actually
	  changes a file
	- darwinup: uninstalling a superseded root does not require
	  the -f option to force it since it is safe in all cases.
	- darwinup: conditionalize some features in case we build
	  without xar or universal support.

Release 16.2 [22-Mar-2010]
	- darwinup: fix crash in verify subcommand
	- darwinup: fix non-root list subcommand locking

Release 16.1 [15-Mar-2010]
	- darwinup: uninstall a root's files in reverse order so children
	  get removed before parents. Fixes "!" status bug.
	- darwinup: lock depot for entire duration of operations to prevent
	  collisions.
	- darwinup: gracefully warn when update_dyld_shared_cache not found
	- darwinup: fix 32-bit-only crash in database subsystem
	- darwinup: adjust what output goes to stderr vs stdout

Release 16 [12-Mar-2010]
	- darwinup: delete old roots after they are no longer needed
	- darwinup: track base OS build for roots
	- darwinup: list subcommand now handles arbitrary arguments
	- darwinup: added superseded keyword to allow cleanup of superseded roots
	- darwinup: improve messages when not running as root

Release 15 [03-Mar-2010]
	- darwinup: fixed stray directories bugs due to apparent local modifications
	- darwinup: added new database abstraction layer to simplify Depot code,
	  reduce copy-paste, and reduce unnecessary statement preparation
	- darwinup: allow list subcommand to run as not-root

darwinup/main.cpp
/*
 * Copyright (c) 2005-2010 Apple Computer, Inc. All rights reserved.
 *
 * @APPLE_BSD_LICENSE_HEADER_START@
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 *
 * 1.  Redistributions of source code must retain the above copyright
 *     notice, this list of conditions and the following disclaimer.
 * 2.  Redistributions in binary form must reproduce the above copyright
 *     notice, this list of conditions and the following disclaimer in the
 *     documentation and/or other materials provided with the distribution.
 * 3.  Neither the name of Apple Computer, Inc. ("Apple") nor the names of
 *     its contributors may be used to endorse or promote products derived
 *     from this software without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY APPLE AND ITS CONTRIBUTORS "AS IS" AND ANY
 * EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
 * WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
 * DISCLAIMED. IN NO EVENT SHALL APPLE OR ITS CONTRIBUTORS BE LIABLE FOR ANY
 * DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
 * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
 * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
 * ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
 * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF
 * THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 *
 * @APPLE_BSD_LICENSE_HEADER_END@
 */

#include <Availability.h>
#include <libgen.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <sys/time.h>
#include <unistd.h>
#include <limits.h>

#include "Archive.h"
#include "Depot.h"
#include "Utils.h"
#include "DB.h"


void usage(char* progname) {
	fprintf(stderr, "usage:    %s [-v] [-p DIR] [command] [args]          \n", progname);
	fprintf(stderr, "version: 33                                                    \n");
	fprintf(stderr, "                                                               \n");
	fprintf(stderr, "options:                                                       \n");
#if __MAC_OS_X_VERSION_MIN_REQUIRED >= 1060
	fprintf(stderr, "          -d        disable helpful automation                 \n");	
#endif
	fprintf(stderr, "          -f        force operation to succeed at all costs    \n");
	fprintf(stderr, "          -n        dry run                                    \n");
	fprintf(stderr, "          -p DIR    operate on roots under DIR (default: /)    \n");
#if __MAC_OS_X_VERSION_MIN_REQUIRED >= 1060
	fprintf(stderr, "          -r        gracefully restart when finished           \n");	
#endif
	fprintf(stderr, "          -v        verbose (use -vv for extra verbosity)      \n");
	fprintf(stderr, "                                                               \n");
	fprintf(stderr, "commands:                                                      \n");
	fprintf(stderr, "          files      <archive>                                 \n");
	fprintf(stderr, "          install    <path>                                    \n");
	fprintf(stderr, "          list       [archive]                                 \n");
	fprintf(stderr, "          rename     <archive> <name>                          \n");
	fprintf(stderr, "          uninstall  <archive>                                 \n");
	fprintf(stderr, "          upgrade    <path>                                    \n");
	fprintf(stderr, "          verify     <archive>                                 \n");
	fprintf(stderr, "                                                               \n");
	fprintf(stderr, "<path> is one of:                                              \n");
	fprintf(stderr, "          /path/to/local/dir-or-file                           \n");
	fprintf(stderr, "          user@host:/path/to/remote/dir-or-file                \n");
	fprintf(stderr, "          http[s]://host/path/to/remote/file                   \n");
	fprintf(stderr, "                                                               \n");
	fprintf(stderr, "Files must be in one of the supported archive formats:         \n");
	fprintf(stderr, "          cpio, cpio.gz, cpio.bz2                              \n");
	fprintf(stderr, "          pax, pax.gz, pax.bz2                                 \n");
	fprintf(stderr, "          tar, tar.gz, tar.bz2                                 \n");
#if __MAC_OS_X_VERSION_MIN_REQUIRED >= 1060	
	fprintf(stderr, "          xar, zip                                             \n");
#else
	fprintf(stderr, "          zip                                                  \n");	
#endif
	fprintf(stderr, "                                                               \n");
	fprintf(stderr, "archive is one of:                                             \n");
	fprintf(stderr, "          <serial>     the Serial number                       \n");
	fprintf(stderr, "          <uuid>       the UUID                                \n");
	fprintf(stderr, "          <name>       the last root installed with that name  \n");
	fprintf(stderr, "          newest       the newest (last) root installed        \n");
	fprintf(stderr, "          oldest       the oldest root installed               \n");
	fprintf(stderr, "          superseded   all roots that have been fully replaced \n");
	fprintf(stderr, "                        by newer roots                         \n");
	fprintf(stderr, "          all          all installed roots                     \n");
	fprintf(stderr, "                                                               \n");
	exit(1);
}

// our globals
uint32_t verbosity;
uint32_t force;
uint32_t dryrun;


int main(int argc, char* argv[]) {
	char* progname = strdup(basename(argv[0]));      
	char* path = NULL;
	bool disable_automation = false;
#if __MAC_OS_X_VERSION_MIN_REQUIRED >= 1060
	bool restart = false;
#endif
	
	int ch;
#if __MAC_OS_X_VERSION_MIN_REQUIRED >= 1060
	while ((ch = getopt(argc, argv, "dfnp:rvh")) != -1) {
#else
	while ((ch = getopt(argc, argv, "dfnp:vh")) != -1) {
#endif
		switch (ch) {
		case 'd':
				disable_automation = true;
				break;
		case 'f':
				force = 1;
				break;
		case 'n':
				dryrun = 1;
				disable_automation = true;
				break;
		case 'p':
				if (optarg[0] != '/') {
					fprintf(stderr, "Error: -p option must be an absolute path\n");
					exit(4);
				}
				if (strlen(optarg) > (PATH_MAX - 1)) {
					fprintf(stderr, "Error: -p option value is too long \n");
					exit(4);
				}
				join_path(&path, optarg, "/");
				break;
#if __MAC_OS_X_VERSION_MIN_REQUIRED >= 1060	
		case 'r':
				restart = true;
				break;
#endif
		case 'v':
				verbosity <<= 1;
				verbosity |= VERBOSE;
				break;
		case '?':
		case 'h':
		default:
				usage(progname);
		}
	}
	argc -= optind;
    argv += optind;
	if (argc == 0) usage(progname);
	
	int res = 0;

	if (dryrun) IF_DEBUG("option: dry run\n");
	if (force)  IF_DEBUG("option: forcing operations\n");
	if (disable_automation) IF_DEBUG("option: helpful automation disabled\n");
#if __MAC_OS_X_VERSION_MIN_REQUIRED >= 1060
    if (restart) IF_DEBUG("option: restart when finished\n");
#endif
	
	if (!path) {
		asprintf(&path, "/");
	} else {
		IF_DEBUG("option: path is %s\n", path);
	}

	Depot* depot = new Depot(path);
		
	// list handles args optional and in special ways
	if (strcmp(argv[0], "list") == 0) {
		res = depot->initialize(false);
		if (res == DEPOT_NOT_EXIST) {
			// we are not asking to write, 
			// but no depot exists yet either,
			// so print an empty list
			depot->archive_header();
			exit(0);
		}
		if (res == DEPOT_PERM_DENIED) {
			// permission denied when trying to read
			// the depot
			fprintf(stderr, "Permission denied when trying to read the database.\n");
			exit(6);
		}
		if (res == 0) depot->list(argc-1, (char**)(argv+1));
	} else if (argc == 1) {
		// other commands which take no arguments
		if (strcmp(argv[0], "dump") == 0) {
			if (depot->initialize(false)) exit(11);
			depot->dump();
		} else {
			fprintf(stderr, "Error: unknown command: '%s' \n", argv[0]);
			usage(progname);
		}
	} else {
		// loop over arguments
		for (int i = 1; i < argc && res == 0; i++) {
			if (strcmp(argv[0], "install") == 0) {
				if (i==1 && depot->initialize(true)) exit(13);
				// gaurd against installing paths ontop of themselves
				if (strncmp(path, argv[i], strlen(argv[i])) == 0 
					&& (strlen(path) == strlen(argv[i]) 
						|| strlen(path) - 1 == strlen(argv[i]))) {
					if (strncmp(path, "/", 1) == 0 && strlen(path) == 1) {
						fprintf(stderr, "Error: You provided '/' as a path to a root. "
								"If you meant to specify a destination of '/', then you "
								"just need to remove the '/' argument as the destination "
								"defaults to '/'. Use the -p option to specify another "
								"destination.\n");
					} else {
						fprintf(stderr, "Error: You cannot install the root at '%s' onto"
								" itself.\n", path);
					}
					res = DEPOT_ERROR;
				}							
				if (res == 0) res = depot->install(argv[i]);
			} else if (strcmp(argv[0], "upgrade") == 0) {
				if (i==1 && depot->initialize(true)) exit(14);
				// find most recent matching archive by name
				Archive* old = depot->get_archive(basename(argv[i]));
				if (!old) {
					fprintf(stderr, "Error: unable to find a matching root to upgrade.\n");
					res = 5;
				}
				// install new archive
				if (res == 0) res = depot->install(argv[i]);
				// uninstall old archive
				if (res == 0) res = depot->uninstall(old);
			} else if (strcmp(argv[0], "files") == 0) {
				if (i==1 && depot->initialize(false)) exit(12);
				res = depot->process_archive(argv[0], argv[i]);
			} else if (strcmp(argv[0], "uninstall") == 0) {
				if (i==1 && depot->initialize(true)) exit(15);
				res = depot->process_archive(argv[0], argv[i]);
			} else if (strcmp(argv[0], "verify") == 0) {
				if (i==1 && depot->initialize(true)) exit(16);
				res = depot->process_archive(argv[0], argv[i]);
			} else if (strcmp(argv[0], "rename") == 0) {
				if (i==1 && depot->initialize(true)) exit(17);
				if ((i+1) >= argc) {
					fprintf(stderr, 
							"Error: rename command for '%s' takes 2 arguments.\n", 
							argv[i]);
					exit(18);
				}
				res = depot->rename_archive(argv[i], argv[i+1]);
				i++;
			} else {
				fprintf(stderr, "Error: unknown command: '%s' \n", argv[0]);
				usage(progname);
			}
		}
#if __MAC_OS_X_VERSION_MIN_REQUIRED >= 1060
		if (!disable_automation && depot->is_dirty() && res == 0) {
			res = update_dyld_shared_cache(path);
			if (res) fprintf(stderr, "Warning: could not update dyld cache.\n");
			res = 0;
		}
		if (!disable_automation && depot->has_modified_extensions() && res == 0) {
			char *sle_path;
			res = join_path(&sle_path, depot->prefix(), "/System/Library/Extensions");
			IF_DEBUG("Touching /System/Library/Extensions\n");
			if (res == 0) res = utimes(sle_path, NULL);
			if (res) {
				fprintf(stderr, "Warning: unable to touch %s \n", sle_path);
				res = 0;
			}
			free(sle_path);
		}
#endif
		if (!disable_automation && depot->has_modified_xpc_services() && res == 0) {
			res = update_xpc_services_cache(path);
			if (res) fprintf(stderr, "Warning: could not update xpc services cache.\n");
			res = 0;
		}
#if __MAC_OS_X_VERSION_MIN_REQUIRED >= 1060
		if (restart && res == 0) {
			res = tell_finder_to_restart();
			if (res) fprintf(stderr, "Warning: tried to tell Finder to restart"
							         "but failed.\n");
			res = 0;
		}
#endif
	}
	
	free(path);
	exit(res);
	return res;
}
